pkg/controller/mssql/grant/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	// enforced by the ProviderConfig schema.
+	ref := pc.Spec.Credentials.ConnectionSecretRef
+	if ref == nil {
+		return nil, errors.New(errNoSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errGetSecret)
+	}
+
+	return &external{
+		db:   c.newDB(s.Data, stringValue(cr.Spec.ForProvider.Database)),
+		kube: c.kube,
+	}, nil
+}
+
+type external struct {
+	db   xsql.DB
+	kube client.Client
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotGrant)
+	}
+
+	permissions, err := c.getPermissions(ctx, *cr.Spec.ForProvider.User)
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	if len(permissions) == 0 {
+		return managed.ExternalObservation{}, nil
+	}
+
+	cr.SetConditions(xpv1.Available())
+
+	g, r := diffPermissions(cr.Spec.ForProvider.Permissions.ToStringSlice(), permissions)
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: len(g) == 0 && len(r) == 0,
+	}, nil
+}
+
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	permissions := strings.Join(cr.Spec.ForProvider.Permissions.ToStringSlice(), ", ")
+
+	query := fmt.Sprintf("GRANT %s TO %s", permissions, mssql.QuoteIdentifier(username))
+	return managed.ExternalCreation{}, errors.Wrap(c.db.Exec(ctx, xsql.Query{String: query}), errGrant)
+}
+
+func (c *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotGrant)
+	}
+
+	desired := cr.Spec.ForProvider.Permissions.ToStringSlice()
+	observed, err := c.getPermissions(ctx, *cr.Spec.ForProvider.User)
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+	toGrant, toRevoke := diffPermissions(desired, observed)
+
+	if len(toRevoke) > 0 {
+		query := fmt.Sprintf("REVOKE %s FROM %s",
+			strings.Join(toRevoke, ", "), mssql.QuoteIdentifier(*cr.Spec.ForProvider.User))
+		if err = c.db.Exec(ctx, xsql.Query{String: query}); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, errRevoke)
+		}
+	}
+	if len(toGrant) > 0 {
+		query := fmt.Sprintf("GRANT %s TO %s",
+			strings.Join(toGrant, ", "), mssql.QuoteIdentifier(*cr.Spec.ForProvider.User))
+		if err = c.db.Exec(ctx, xsql.Query{String: query}); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, errGrant)
+		}
+	}
+	return managed.ExternalUpdate{}, nil
+}
+
+func (c *external) Delete(ctx context.Context, mg resource.Managed) error {
```

Comment:
nit: I did not check other implementations but I think we could wire this logic to some sort of deletion policy, which allows the permissions not to be revoked when the MR is deleted. But if other implementation also behave like this, I think we should be fine revoking these permissions.

Reply:
I believe that could already be possible with deletionPolicy=orphan.